File: content/examples/other/leetcode/problems/reorder-list/example-2/solution.js
class ListNode {
  constructor(val, next = null) {
    this.val = val;
    this.next = next;
  }
}

function reorderList(head) {
  if (!head || !head.next) return head;
  // 1) find middle
  let slow = head;
  let fast = head;
  while (fast && fast.next) {
    slow = slow.next;
    fast = fast.next.next;
  }
  // 2) reverse second half starting at slow
  let prev = null;
  let cur = slow;
  while (cur) {
    const nxt = cur.next;
    cur.next = prev;
    prev = cur;
    cur = nxt;
  }
  // 3) merge first half and reversed second half
  let first = head;
  let second = prev;
  while (second && first) {
    const t1 = first.next;
    const t2 = second.next;
    first.next = second;
    if (!t1) break;
    second.next = t1;
    first = t1;
    second = t2;
  }
  // ensure tail termination (for even lengths)
  if (first) first.next = null;
  return head;
}

function toArray(head) {
  const out = [];
  for (let cur = head; cur; cur = cur.next) out.push(cur.val);
  return out;
}

if (require.main === module) {
  const a = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4))));
  reorderList(a);
  console.log(toArray(a));
}

module.exports = { ListNode, reorderList };

File: content/examples/other/leetcode/problems/reorder-list/example-2/solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def reorder_list(head):
    if head is None or head.next is None:
        return head

    slow = head
    fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next

    prev = None
    cur = slow
    while cur:
        nxt = cur.next
        cur.next = prev
        prev = cur
        cur = nxt

    first = head
    second = prev
    while second and first:
        t1 = first.next
        t2 = second.next
        first.next = second
        if t1 is None:
            break
        second.next = t1
        first = t1
        second = t2

    if first:
        first.next = None
    return head


def to_array(head):
    out = []
    cur = head
    while cur:
        out.append(cur.val)
        cur = cur.next
    return out


if __name__ == "__main__":
    a = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
    reorder_list(a)
    print(to_array(a))
